### PR TITLE
Backend operations

### PR DESCRIPTION
**Pull request by nehalist**

### CODE REVIEW

File: Components/Data/CreditCardPaymentConfig.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    public function setThreeDSecret($threeDSecret)
+    {
+        $this->threeDSecret = $threeDSecret;
+    }
+
+    /**
+     * @return string|float
+     */
+    public function getThreeDSslMaxLimit()
+    {
```
Comment (jakubpolomsky) [inline Components/Data/CreditCardPaymentConfig.php]: when does this occur? shouldn't there be separate 3ds and ssl?

File: Components/Mapper/ArrayMapper.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    protected function get($keys)
+    {
+        $keys  = (array)$keys;
+        $value = $this->arrayEntity;
+        foreach ($keys as $key) {
+            if (! is_array($value) || ! isset($value[$key])) {
+                throw new ArrayKeyNotFoundException(implode('.', $keys), get_class($this), $this->arrayEntity);
+            }
+            $value = $value[$key];
```
Comment (jakubpolomsky) [inline Components/Mapper/ArrayMapper.php]: return $value[$key] directly, no need for keeping to loop over the residual array keys

File: Components/Mapper/ArrayMapper.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    protected function has($keys)
+    {
+        $keys  = (array)$keys;
+        $value = $this->arrayEntity;
+        foreach ($keys as $key) {
+            if (! is_array($value) || ! array_key_exists($key, $value)) {
+                return false;
+            }
+            $value = $value[$key];
```
Comment (jakubpolomsky) [inline Components/Mapper/ArrayMapper.php]: unused line, please remove it

File: Components/Mapper/BasketMapper.php [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            $basket->add($basketItem->getWirecardItem());
+        }
+
+        $shippingCosts = $this->getOptional(self::SHIPPING_COSTS_WITH_TAX);
+        if ($shippingCosts) {
+            $shippingAmount = new Amount(self::numberFormat($shippingCosts), $this->currency);
+
+            $shippingTaxValue = $shippingCosts - $this->getOptional(self::SHIPPING_COSTS_NET, 0.0);
+
+            $basketItem = new Item('Shipping', $shippingAmount, 1);
```
Comment (jakubpolomsky) [inline Components/Mapper/BasketMapper.php]: Would it be possible to get this item name translated directly from the shopsystem?

File: Components/Services/ReturnHandler.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+     *
+     * @return ViewAction
+     */
+    protected function handleFormInteraction(FormInteractionResponse $response)
+    {
+        try {
+            $order = $this->getOrderFromResponse($response);
+            $this->transactionFactory->create($order->getNumber(), $response, Transaction::TYPE_RETURN);
+        } catch (OrderNotFoundException $e) {
+            $this->logger->notice('Could not create transaction for FormInteractionResponse: ' . $e->getMessage());
```
Comment (jakubpolomsky) [inline Components/Services/ReturnHandler.php]: This log message doesn't seem right to me.

File: Controllers/Frontend/WirecardElasticEnginePayment.php [hunk 6]
```diff
@@ -32,329 +32,348 @@
...
 
-        $response = null;
-        if ($request['method'] === PaypalPayment::PAYMETHOD_IDENTIFIER) {
-            $paypal = new PaypalPayment();
-            $response = $paypal->getPaymentResponse($request);
-        }
+        $transactionService = new TransactionService($payment->getTransactionConfig(
+            $this->getModelManager()->getRepository(Shop::class)->getActiveDefault(),
+            $this->container->getParameterBag(),
+            $this->getCurrencyShortName()
```
Comment (jakubpolomsky) [inline Controllers/Frontend/WirecardElasticEnginePayment.php]: Setting the LoggerInterface would save you having to log requests / responses manually. Consider using this approach.

File: Components/Mapper/ArrayMapper.php (hunk 2 above)
Comment (danleeb) [reply Components/Mapper/ArrayMapper.php]: This is to access a multidimensional array, e.g. https://github.com/wirecard/shopware-ee/blob/backend-operations/Components/Mapper/UserMapper.php#L177
Or am I missing something here?

File: Components/Mapper/ArrayMapper.php (hunk 3 above)
Comment (danleeb) [reply Components/Mapper/ArrayMapper.php]: Used to check all keys in a multidimensional array. I will add a comment to clarify this behavious.

File: Controllers/Frontend/WirecardElasticEnginePayment.php (hunk 6 above)
Comment (danleeb) [reply Controllers/Frontend/WirecardElasticEnginePayment.php]: We missed this one, thank you. Fixed

File: Exception/NotAvailableBasketException.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+namespace WirecardShopwareElasticEngine\Exception;
+
+use WirecardShopwareElasticEngine\Components\Mapper\BasketItemMapper;
+use WirecardShopwareElasticEngine\Components\Mapper\BasketMapper;
+
+class NotAvailableBasketException extends BasketException
+{
+    public function __construct(array $article, BasketItemMapper $item, BasketMapper $basket)
+    {
+        parent::__construct("Article '{$item->getArticleName()}' ({$item->getArticleNumber()}) is not available");
```
[the remark this answers is not in the document]
Comment (jakubpolomsky) [reply Exception/NotAvailableBasketException.php]: same in ordernotfoundexception and outofstockbasketexception

[the remark this answers is not in the document]
Comment (jakubpolomsky) [reply Exception/NotAvailableBasketException.php]: that is what i meant, sorry

File: Components/Mapper/ArrayMapper.php (hunk 2 above)
Comment (jakubpolomsky) [reply Components/Mapper/ArrayMapper.php]: ok

File: Components/Mapper/ArrayMapper.php (hunk 3 above)
Comment (jakubpolomsky) [reply Components/Mapper/ArrayMapper.php]: As it is not clear that the keys in the argument are always in the correct order (without looking at the mappers) a comment would really be appreciated. Thank you!

**Pull request merged**